gtk/Session.h

```diff
@@ -158,5 +158,5 @@ class Session : public Glib::Object
 
 private:
     class Impl;
-    std::unique_ptr<Impl> const impl_;
+    std::shared_ptr<Impl> const impl_;
```

Comment:
Please revert. If you want to avoid using `Session` instance past its destruction, store a `sigc::connection` and disconnect in destructor so that callback isn't called at all; see `Session::Impl::monitor_idle_tag_` as one of examples. If multiple removal operations could be started concurrently, container of connections (map torrent id -> connection?) will do too.

Reply:
Sorry, it's late and I'm tired, the above makes little sense. Still, switching to `shared_ptr` for `Impl` is a no-go, I'll think of something else (or you might). One obvious choice is `Impl::get_core_ptr()` + `Glib::WeakRef`, but maybe there's a better way...

Reply:
Oh, `Session` is already reference counted, I missed that. I found no `Glib::WeakRef`, but I guess there is no problem in holding a strong reference until the delete function answers.